schema/spacewalk/upgrade/susemanager-schema-4.2.2-to-susemanager-schema-4.2.3/1000-debvercmp.sql

```diff
@@ -0,0 +1,506 @@
+        on conflict do nothing;
+
+    select id
+        into strict evr_id
+        from rhnPackageEVR
+        where ((epoch is null and e_in is null) or (epoch = e_in)) and
+           version = v_in and release = r_in and (evr).type = t_in;
+
+    return evr_id;
+end;
+$$ language plpgsql;
+
+
+-- update lookup_evr
+create or replace function
+lookup_evr(e_in in varchar, v_in in varchar, r_in in varchar, t_in in varchar)
+returns numeric
+as
+$$
+declare
+    evr_id  numeric;
+begin
+    select id
+      into evr_id
+      from rhnPackageEVR
+     where ((epoch is null and e_in is null) or (epoch = e_in)) and
+           version = v_in and
+           release = r_in and
+           (evr).type = t_in;
+
+    if not found then
+        -- HACK: insert is isolated in own function in order to be able to declare this function immutable
+        -- Postgres optimizes immutable functions calls but those are compatible with the contract of lookup_\*
+        -- see https://www.postgresql.org/docs/9.6/xfunc-volatility.html
+        return insert_evr(e_in, v_in, r_in, t_in);
+    end if;
+
+    return evr_id;
+end;
+$$ language plpgsql immutable;
+
+
+create or replace function
+lookup_transaction_package(
+    o_in in varchar,
+    n_in in varchar,
+    e_in in varchar,
+    v_in in varchar,
+    r_in in varchar,
+    a_in in varchar)
+returns numeric
+as
+$$
+declare
+    o_id        numeric;
+    n_id        numeric;
+    e_id        numeric;
+    p_arch_id   numeric;
+    tp_id       numeric;
+    type        varchar;
+begin
+    select id
+      into o_id
+      from rhnTransactionOperation
+     where label = o_in;
+
+    if not found then
+        perform rhn_exception.raise_exception('invalid_transaction_operation');
+    end if;
+
+    n_id := lookup_package_name(n_in);
+    p_arch_id := null;
+
+    if a_in is not null then
+        p_arch_id := lookup_package_arch(a_in);
+        SELECT t.label into type from rhnpackagearch pa join rhnarchtype t
+         on t.id = pa.arch_type_id where pa.id = p_arch_id;
+    end if;
+
+    e_id := lookup_evr(e_in, v_in, r_in, type);
+
+    select id
+      into tp_id
+      from rhnTransactionPackage
+     where operation = o_id and
+           name_id = n_id and
+           evr_id = e_id and
+           (package_arch_id = p_arch_id or (p_arch_id is null and package_arch_id is null));
+
+    if not found then
+        -- HACK: insert is isolated in own function in order to be able to declare this function immutable
+        -- Postgres optimizes immutable functions calls but those are compatible with the contract of lookup_\*
+        -- see https://www.postgresql.org/docs/9.6/xfunc-volatility.html
+        return insert_transaction_package(o_id, n_id, e_id, p_arch_id);
+    end if;
+    return tp_id;
+end;
+$$ language plpgsql immutable;
+
+
+ALTER TABLE rhnpackageevr DISABLE TRIGGER USER;
+
+-- set all existing rhnpackageevr to rpm
+update rhnpackageevr set evr.type = 'rpm';
```

Comment:
I think this will create a mess when we run it multiple times.

```suggestion
update rhnpackageevr set evr.type = 'rpm' where evr.type is NULL;
```